File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/domain/events/GiftCardRedeemEvents.java
package it.giannibombelli.wsc2026.giftcard.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;

/**
 * Sealed interface so the aggregate can return exactly one of the two possible outcomes.
 */
public sealed interface GiftCardRedeemEvents extends GiftCardEvent
    permits GiftCardRedeemEvents.GiftCardRedeemed, GiftCardRedeemEvents.GiftCardNotRedeemed {

    record GiftCardRedeemed(
        GiftCardId aggregateId,
        Money redeemedAmount
    ) implements GiftCardRedeemEvents {

        public GiftCardRedeemed {
            Require.requireArgument(aggregateId, "giftCardId");
            Require.requireArgument(redeemedAmount, "redeemedAmount");
        }
    }

    record GiftCardNotRedeemed(
        GiftCardId aggregateId,
        Money attemptedAmount,
        Description reason
    ) implements GiftCardRedeemEvents {

        public GiftCardNotRedeemed {
            Require.requireArgument(aggregateId, "giftCardId");
            Require.requireArgument(attemptedAmount, "attemptedAmount");
            Require.requireArgument(reason, "reason");
        }
    }
}

File: typescript/src/giftcard/domain/events/giftCardRedeemEvents.ts
// Eventi emessi quando un riscatto della gift card riesce o fallisce.

import { Description } from '@/common/domain/primitive/description.js';
import type { Event } from '@/common/domain/model/event.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { GiftCardId } from '../giftcard/giftCardId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type GiftCardRedeemed = Event<'GiftCardRedeemed', GiftCardId> & {
  readonly redeemedAmount: Money;
};

export function giftCardRedeemed(aggregateId: GiftCardId, redeemedAmount: Money): GiftCardRedeemed {
  requireArgument(aggregateId, 'giftCardId');
  requireArgument(redeemedAmount, 'redeemedAmount');
  return { kind: 'GiftCardRedeemed', aggregateId, redeemedAmount };
}

export type GiftCardNotRedeemed = Event<'GiftCardNotRedeemed', GiftCardId> & {
  readonly attemptedAmount: Money;
  readonly reason: Description;
};

export function giftCardNotRedeemed(
  aggregateId: GiftCardId,
  attemptedAmount: Money,
  reason: Description
): GiftCardNotRedeemed {
  requireArgument(aggregateId, 'giftCardId');
  requireArgument(attemptedAmount, 'attemptedAmount');
  requireArgument(reason, 'reason');
  return { kind: 'GiftCardNotRedeemed', aggregateId, attemptedAmount, reason };
}

export type GiftCardRedeemEvent = GiftCardRedeemed | GiftCardNotRedeemed;
